In [1]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
from credentials import user
from credentials import server
from credentials import password

In [2]:
connString = "postgres+psycopg2://"+user+":"+password+"@"+server+":5432/covid_data"
engine = create_engine(connString)

In [3]:
mask_use_df = pd.read_sql("SELECT * FROM mask_use;", con=connString)
counties_df = pd.read_sql("SELECT * FROM counties;", con=connString)

In [4]:
pop_df = pd.read_csv("../data/co-est2019-alldata.csv",encoding='latin-1')

In [5]:
pop_df.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2019  \
0        4779736            4780125          4785437  ...          1.917501   
1          54571              54597            54773  ...          4.847310   
2         182265             182265           183112  ...         24.017829   
3          27457              27455            27327  ...         -5.690302   
4          22915              22915            22870  ...          1.385134   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  \
0     0.578434     1.186314     1.522549     0.563489     0.626357   
1     6.018182    -6.226119    -3.902226     1.970443    -1.712875   
2    16.641870    17.488579    22.751474    20.184334    17.725964   
3     0.292676    -6.897817    -8.132185    -5.140431   -15.724575   
4    -4.998356    -3.787545    -5.797999     1.331144     1.329817   

   RNETMIG2016  RNETMIG2017  RNETMIG2018  RNETMIG2019  
0     0.745172     1.090366     1.773786     2.483744  
1     4.777171     0.849656     0.540916     4.560062  
2    21.279291    22.398256    24.727215    24.380567  
3   -18.238016   -24.998528    -8.754922    -5.165664  
4    -0.708717    -3.234669    -6.857092     1.831952  

[5 rows x 164 columns]

In [6]:
pop_df = pop_df[['STNAME','CTYNAME','POPESTIMATE2019']]

In [7]:
pop_df.columns = ['state','county','pop_est_2019']
pop_df = pop_df.query("state != county")
pop_df['county'] = pop_df['county'].str.replace(' County', '')
pop_df['county'] = pop_df['county'].str.replace(' Parish', '')
pop_df.head()

state   county  pop_est_2019
1  Alabama  Autauga         55869
2  Alabama  Baldwin        223234
3  Alabama  Barbour         24686
4  Alabama     Bibb         22394
5  Alabama   Blount         57826

In [8]:
mask_use_df.head()

county_fips  never  rarely  sometimes  frequently  always
0       01001  0.053   0.074      0.134       0.295   0.444
1       01003  0.083   0.059      0.098       0.323   0.436
2       01005  0.067   0.121      0.120       0.201   0.491
3       01007  0.020   0.034      0.096       0.278   0.572
4       01009  0.053   0.114      0.180       0.194   0.459

In [9]:
counties_df.head()

date     county       state   fips  cases  deaths
0  2020-01-21  Snohomish  Washington  53061      1       0
1  2020-01-22  Snohomish  Washington  53061      1       0
2  2020-01-23  Snohomish  Washington  53061      1       0
3  2020-01-24       Cook    Illinois  17031      1       0
4  2020-01-24  Snohomish  Washington  53061      1       0

In [10]:
counties_df.columns = ['date','county','state','county_fips','cases','deaths']

In [11]:
mask_counties = pd.merge(mask_use_df, counties_df, on='county_fips')
mask_counties = pd.merge(mask_counties, pop_df, on=['county','state'], how='left')

In [12]:
mask_counties.head()

county_fips  never  rarely  sometimes  frequently  always        date  \
0       01001  0.053   0.074      0.134       0.295   0.444  2020-03-24   
1       01001  0.053   0.074      0.134       0.295   0.444  2020-03-25   
2       01001  0.053   0.074      0.134       0.295   0.444  2020-03-26   
3       01001  0.053   0.074      0.134       0.295   0.444  2020-03-27   
4       01001  0.053   0.074      0.134       0.295   0.444  2020-03-28   

    county    state  cases  deaths  pop_est_2019  
0  Autauga  Alabama      1       0       55869.0  
1  Autauga  Alabama      4       0       55869.0  
2  Autauga  Alabama      6       0       55869.0  
3  Autauga  Alabama      6       0       55869.0  
4  Autauga  Alabama      6       0       55869.0

In [13]:
mask_counties = mask_counties.drop_duplicates(subset=['county_fips'])
mask_counties.head()

county_fips  never  rarely  sometimes  frequently  always        date  \
0         01001  0.053   0.074      0.134       0.295   0.444  2020-03-24   
238       01003  0.083   0.059      0.098       0.323   0.436  2020-03-14   
486       01005  0.067   0.121      0.120       0.201   0.491  2020-04-03   
714       01007  0.020   0.034      0.096       0.278   0.572  2020-03-30   
946       01009  0.053   0.114      0.180       0.194   0.459  2020-03-25   

      county    state  cases  deaths  pop_est_2019  
0    Autauga  Alabama      1       0       55869.0  
238  Baldwin  Alabama      1       0      223234.0  
486  Barbour  Alabama      1       0       24686.0  
714     Bibb  Alabama      2       0       22394.0  
946   Blount  Alabama      1       0       57826.0

In [14]:
good_compliance = []
>>> for index, row in mask_counties.iterrows():
...     good_compliance.append(row.frequently + row.always)

In [15]:
mask_counties['good_compliance'] = good_compliance

In [16]:
mask_counties = mask_counties.reset_index(drop=True)
mask_counties['good_rank'] = mask_counties['good_compliance'].rank(ascending=False)
mask_counties.tail()

county_fips  never  rarely  sometimes  frequently  always        date  \
3128       56037  0.061   0.295      0.230       0.146   0.268  2020-03-24   
3129       56039  0.095   0.157      0.160       0.247   0.340  2020-03-18   
3130       56041  0.098   0.278      0.154       0.207   0.264  2020-04-02   
3131       56043  0.204   0.155      0.069       0.285   0.287  2020-03-27   
3132       56045  0.142   0.129      0.148       0.207   0.374  2020-05-27   

          county    state  cases  deaths  pop_est_2019  good_compliance  \
3128  Sweetwater  Wyoming      1       0       42343.0            0.414   
3129       Teton  Wyoming      1       0       23464.0            0.587   
3130       Uinta  Wyoming      1       0       20226.0            0.471   
3131    Washakie  Wyoming      1       0        7805.0            0.572   
3132      Weston  Wyoming      1       0        6927.0            0.581   

      good_rank  
3128     3107.0  
3129     2581.0  
3130     3022.5  
3131     2654.0  
3132     2615.0

In [17]:
davidson = mask_counties.loc[mask_counties['county'] == 'Davidson'].reset_index(drop=True)
davidson = davidson.loc[davidson['state'] == 'Tennessee'].reset_index(drop=True)

In [18]:
davidson.head()

county_fips  never  rarely  sometimes  frequently  always        date  \
0       47037   0.02   0.037      0.086       0.181   0.677  2020-03-08   

     county      state  cases  deaths  pop_est_2019  good_compliance  \
0  Davidson  Tennessee      1       0      694144.0            0.858   

   good_rank  
0      497.5

In [19]:
tennessee = mask_counties.loc[mask_counties['state'] == 'Tennessee'].reset_index(drop=True)
tennessee['good_rank'] = tennessee['good_compliance'].rank(ascending=False)
tennessee.sort_values('good_compliance', ascending=False)

county_fips  never  rarely  sometimes  frequently  always        date  \
23       47047  0.022   0.001      0.089       0.205   0.682  2020-03-21   
18       47037  0.020   0.037      0.086       0.181   0.677  2020-03-08   
78       47157  0.034   0.025      0.105       0.184   0.653  2020-03-08   
93       47187  0.019   0.073      0.091       0.212   0.604  2020-03-05   
85       47171  0.049   0.041      0.108       0.306   0.497  2020-03-26   
..         ...    ...     ...        ...         ...     ...         ...   
43       47087  0.138   0.128      0.249       0.093   0.392  2020-04-02   
70       47141  0.160   0.164      0.207       0.099   0.369  2020-03-21   
7        47015  0.209   0.089      0.238       0.182   0.282  2020-03-26   
20       47041  0.157   0.152      0.231       0.132   0.328  2020-03-24   
92       47185  0.185   0.205      0.162       0.115   0.333  2020-03-25   

        county      state  cases  deaths  pop_est_2019  good_compliance  \
23     Fayette  Tennessee      1       0       41133.0            0.887   
18    Davidson  Tennessee      1       0      694144.0            0.858   
78      Shelby  Tennessee      1       0      937166.0            0.837   
93  Williamson  Tennessee      1       0      238412.0            0.816   
85      Unicoi  Tennessee      1       0       17883.0            0.803   
..         ...        ...    ...     ...           ...              ...   
43     Jackson  Tennessee      2       0       11786.0            0.485   
70      Putnam  Tennessee      5       0       80245.0            0.468   
7       Cannon  Tennessee      1       0       14678.0            0.464   
20      DeKalb  Tennessee      1       0       20490.0            0.460   
92       White  Tennessee      1       0       27345.0            0.448   

    good_rank  
23        1.0  
18        2.0  
78        3.0  
93        4.0  
85        5.0  
..        ...  
43       91.0  
70       92.0  
7        93.0  
20       94.0  
92       95.0  

[95 rows x 14 columns]

In Tennessee, Davidson County has the second highest percentage of respondents who always or frequently use masks.

In [20]:
tennessee['good_compliance'].mean()

0.6367157894736842

Davidson County runs about 22% above the Tennessee mean in frequent or better mask usage...

In [21]:
mask_counties['good_compliance'].mean()

0.7153916374082333

...and 14% above the US mean, ranking slightly better than 500th out of 3132 counties. Tennessee as a whole is 8% below the US mean.

In [22]:
top_50_counties = mask_counties.nlargest(50,'pop_est_2019').reset_index(drop=True)
top_50_counties = pd.concat([davidson,top_50_counties]).reset_index(drop=True)
top_50_counties['good_rank'] = top_50_counties['good_compliance'].rank(ascending=False)
top_50_counties

county_fips  never  rarely  sometimes  frequently  always        date  \
0        47037  0.020   0.037      0.086       0.181   0.677  2020-03-08   
1        06037  0.021   0.013      0.049       0.131   0.786  2020-01-26   
2        17031  0.023   0.021      0.072       0.162   0.722  2020-01-24   
3        48201  0.019   0.024      0.069       0.152   0.736  2020-03-05   
4        04013  0.023   0.025      0.059       0.158   0.734  2020-01-26   
5        06073  0.017   0.023      0.034       0.126   0.800  2020-02-10   
6        06059  0.023   0.021      0.046       0.156   0.754  2020-01-25   
7        12086  0.032   0.023      0.060       0.128   0.756  2020-03-11   
8        48113  0.024   0.019      0.059       0.141   0.757  2020-03-10   
9        06065  0.026   0.014      0.041       0.116   0.803  2020-03-07   
10       32003  0.027   0.032      0.054       0.145   0.742  2020-03-05   
11       53033  0.016   0.020      0.050       0.191   0.724  2020-02-28   
12       06071  0.027   0.022      0.048       0.134   0.768  2020-03-15   
13       48439  0.023   0.034      0.077       0.144   0.722  2020-03-10   
14       48029  0.015   0.016      0.057       0.090   0.822  2020-02-12   
15       12011  0.026   0.023      0.043       0.117   0.791  2020-03-06   
16       06085  0.015   0.014      0.040       0.168   0.764  2020-01-31   
17       26163  0.046   0.032      0.075       0.156   0.690  2020-03-10   
18       06001  0.019   0.008      0.055       0.123   0.795  2020-03-01   
19       25017  0.023   0.011      0.041       0.127   0.798  2020-03-05   
20       42101  0.023   0.012      0.051       0.120   0.794  2020-03-10   
21       06067  0.025   0.028      0.062       0.164   0.722  2020-02-21   
22       12099  0.030   0.020      0.050       0.116   0.784  2020-03-12   
23       36103  0.040   0.019      0.053       0.106   0.782  2020-03-08   
24       12057  0.023   0.031      0.069       0.155   0.722  2020-03-01   
25       12095  0.020   0.020      0.072       0.164   0.724  2020-03-13   
26       36059  0.017   0.018      0.043       0.142   0.779  2020-03-05   
27       39049  0.042   0.054      0.090       0.198   0.617  2020-03-14   
28       48453  0.015   0.008      0.024       0.158   0.795  2020-03-13   
29       27053  0.021   0.044      0.096       0.243   0.596  2020-03-12   
30       26125  0.028   0.027      0.049       0.169   0.727  2020-03-10   
31       39035  0.042   0.052      0.084       0.213   0.609  2020-03-09   
32       42003  0.017   0.022      0.035       0.165   0.761  2020-03-14   
33       49035  0.028   0.032      0.094       0.202   0.644  2020-02-25   
34       06013  0.018   0.016      0.039       0.121   0.806  2020-03-03   
35       51059  0.018   0.016      0.054       0.163   0.748  2020-03-07   
36       37183  0.015   0.020      0.070       0.168   0.727  2020-03-03   
37       37119  0.029   0.037      0.075       0.166   0.694  2020-03-12   
38       13121  0.026   0.036      0.089       0.187   0.662  2020-03-02   
39       24031  0.025   0.029      0.041       0.128   0.777  2020-03-05   
40       04019  0.023   0.024      0.073       0.121   0.759  2020-03-09   
41       48085  0.021   0.026      0.054       0.181   0.719  2020-03-09   
42       06019  0.021   0.022      0.059       0.156   0.741  2020-03-07   
43       29189  0.024   0.028      0.069       0.182   0.697  2020-03-07   
44       12103  0.025   0.019      0.074       0.133   0.750  2020-03-10   
45       15003  0.011   0.012      0.027       0.151   0.798  2020-03-06   
46       36119  0.025   0.018      0.044       0.115   0.798  2020-03-04   
47       18097  0.051   0.053      0.116       0.232   0.548  2020-03-06   
48       12031  0.049   0.042      0.086       0.194   0.629  2020-03-12   
49       55079  0.053   0.067      0.119       0.264   0.497  2020-03-13   
50       09001  0.027   0.019      0.060       0.114   0.780  2020-03-08   

            county           state  cases  deaths 

In [23]:
#top_50_counties.to_csv('../data/top50.csv',index=False)

In [24]:
#tennessee.to_csv('../data/tennessee.csv',index=False)